stl/inc/format

```diff
@@ -1248,15 +1243,449 @@ _OutputIt _Write(_OutputIt _Out, basic_string_view<_CharT> _Value) {
+        }
+        _Format = chars_format::fixed;
+        break;
+    case 'G':
+        _To_upper = true;
+        [[fallthrough]];
+    case 'g':
+        if (_Specs._Precision == -1) {
+            _Specs._Precision = 6;
+        }
+        _Format   = chars_format::general;
+        _Exponent = 'e';
+        break;
+    default:
+        throw format_error("invalid floating point type");
+    }
+
+    array<char, _Format_min_buffer_length> _Buffer;
+    to_chars_result _Result;
+
+    if (_Specs._Precision == -1) {
+        _Result = to_chars(_Buffer.data(), _Buffer.data() + _Buffer.size(), _Value, _Format);
+    } else {
+        _Result = to_chars(_Buffer.data(), _Buffer.data() + _Buffer.size(), _Value, _Format, _Specs._Precision);
+    }
+
+    _STL_ASSERT(_Result.ec == errc{}, "to_chars failed");
+
+    auto _Buffer_start = _Buffer.data();
+    auto _Width        = static_cast<int>(_Result.ptr - _Buffer_start);
+
+    const auto _Is_negative = signbit(_Value);
+
+    if (_Is_negative) {
+        // Remove the '-', it will be dealt with directly
+        _Buffer_start += 1;
+    } else {
+        if (_Specs._Sgn != _Sign::_Minus) {
+            _Width += 1;
+        }
+    }
+
+    if (_To_upper) {
+        _Fmt_buffer_to_uppercase(_Buffer_start, _Result.ptr);
+        _Exponent = static_cast<char>(_CSTD toupper(_Exponent));
+    }
+
+    const auto _Is_finite = !_STD isnan(_Value) && !_STD isinf(_Value);
+
+    auto _Append_decimal   = false;
+    auto _Exponent_start   = _Result.ptr;
+    auto _Zeroes_to_append = 0;
+
+    if (_Specs._Alt && _Is_finite) {
+        auto _Radix_point = _Result.ptr;
+        for (auto it = _Buffer_start; it < _Result.ptr; ++it) {
+            if (*it == '.') {
+                _Radix_point = it;
+            } else if (*it == _Exponent) {
+                _Exponent_start = it;
+            }
+        }
+        if (_Radix_point == _Result.ptr) {
+            ++_Width;
+            _Append_decimal = true;
+        }
+
+        if (_Specs._Type == 'g' || _Specs._Type == 'G') {
+            _Zeroes_to_append = _Specs._Precision - static_cast<int>(_Exponent_start - _Buffer_start);
+            if (!_Append_decimal) {
+                _Zeroes_to_append += 1;
+            }
+            _Width += _Zeroes_to_append;
+        }
+    }
+
+    const bool _Write_leading_zeroes = _Specs._Leading_zero && _Specs._Alignment == _Align::_None;
+    auto _Writer                     = [&](_OutputIt _Out) {
+        _Out = _Write_sign(_Out, _Specs._Sgn, _Is_negative);
+
+        if (_Write_leading_zeroes && _Width < _Specs._Width && _Is_finite) {
+            _Out = _STD fill_n(_Out, _Specs._Width - _Width, '0');
+        }
+
+        if (_Specs._Alt) {
+            _Out          = _STD copy(_Buffer_start, _Exponent_start, _Out);
+            _Buffer_start = _Exponent_start;
+
+            if (_Append_decimal) {
+                *_Out++ = '.';
+            }
+            while (_Zeroes_to_append-- > 0) {

```

Comment:
Nitpick: I think this would be clearer as a `for` loop: `for (; _Zeroes_to_append > 0; --_Zeroes_to_append)`. The rationale is that `for` loops are the most highly structured and are the easiest to reason about (unless they to do something wacky, which they shouldn't). `while` loops are less structured and call for more general reasoning. Here this is just a simple countdown, so the `for` loop immediately reads as that, whereas for the `while` loop I have to think about the sequencing of postdecrement.